src/tools/rbd_mirror/ImageMap.h

```diff
@@ -0,0 +1,278 @@
+  };
+
+  // Lock ordering: m_threads->timer_lock, m_lock
+
+  librados::IoCtx &m_ioctx;
+  Threads<ImageCtxT> *m_threads;
+  ImageMapListener &m_listener;
+
+  std::unique_ptr<image_map::Policy<ImageCtxT> > m_policy; // our mapping policy
+
+  Context *m_timer_task;
+  Mutex m_lock;
+  bool m_shutting_down = false;
+  AsyncOpTracker m_async_op_tracker;
+
+  ImageAction m_image_action;               // current and pending actions for images
+  std::map<std::string, Info> m_image_info;
+
+  Updates m_map_updates;
+  Updates m_unmap_updates;
+  std::set<std::string> m_remove_global_image_ids;
+  std::set<std::string> m_global_image_ids;
+
+  std::set<std::string> m_dead_instances;  // list of dead (removed) instances
+
+  struct C_LoadMap : Context {
+    ImageMap *image_map;
+    Context *on_finish;
+
+    std::map<std::string, cls::rbd::MirrorImageMap> image_mapping;
+
+    C_LoadMap(ImageMap *image_map, Context *on_finish)
+      : image_map(image_map),
+        on_finish(on_finish) {
+    }
+
+    void finish(int r) override {
+      if (r == 0) {
+        image_map->handle_load(image_mapping);
+      }
+
+      on_finish->complete(r);
+      image_map->finish_async_op();
+    }
+  };
+
+  struct C_NotifyListener : Context {
+    ImageMap *image_map;
+
+    Updates to_map;
+    std::set<std::string> to_purge;
+
+    C_NotifyListener(ImageMap *image_map, Updates &&to_map,
+                     std::set<std::string> &&to_purge)
+      : image_map(image_map),
+        to_map(to_map),
+        to_purge(to_purge) {
+    }
+
+    void finish(int r) override {
+      // purged images are always successfully completed (update
+      // failure may lead to stale on-disk entries though). note
+      // that state machine transition table (ImageAction) does
+      // not have transition states from "purge" state (except
+      // advancing to "complete").
+      image_map->continue_action(to_purge, 0);
+
+      if (r == 0) {
+        image_map->notify_listener_acquire_release_images(to_map, {});
+      } else {
+        std::set<std::string> retry_global_image_ids;
+
+        for (auto const &update : to_map) {
+          retry_global_image_ids.insert(update.global_image_id);
+        }
+
+        image_map->continue_action(retry_global_image_ids, r);
+      }
+
+      image_map->finish_async_op();
+    }
+  };
+
+  // retry friendly callback contexts for sub actions. note that
+  // callback context when an action is completed need not be
+  // of this type as its just invoked once, upon completion.
+  struct SubActionContext : Context {
+    ImageMap *image_map;
+    std::string global_image_id;
+    std::string instance_id;
+
+    SubActionContext(ImageMap *image_map,
+                     const std::string &global_image_id,
+                     const std::string &instance_id)
+      : image_map(image_map),
+        global_image_id(global_image_id),
+        instance_id(instance_id) {
+    }
+
+    void update_map_sub_action() {
```

Comment:
Nit: why have these "helper" methods instead of just directly invoking the methods from the appropriate ```finish``` override?

Reply:
No strong reason to have it this way... `SubActionContext` invokes `set_retry_instance_id()` which requires it to record `global_image_id` and `instance_id`. Alternatively, each derived class could have accessed the base class member variable (via `SubActionContext::`) and invoked the appropriate methods in `finish()`. Just found this a bit cleaner.

I'll relook at this w.r.t. your comment on integrating `Policy` and `ImageActions`.